File: mercury/dyn_tower.m
%---------------------------------------------------------------------------%
% vim: ft=mercury ts=4 sw=4 et
%---------------------------------------------------------------------------%
% File: dyn_tower.m
% Main author: nyc
%
% Dynamically-nested algebraic extension tower.
%
% Unlike ext_elem which uses static type nesting, tower_elem uses a
% recursive data type supporting towers of arbitrary runtime-determined
% depth.  Elements support full field arithmetic and can be converted
% to radical expressions.
%
%---------------------------------------------------------------------------%

:- module dyn_tower.
:- interface.

:- import_module list.
:- import_module integer.
:- import_module maybe.
:- import_module rad_expr.
:- import_module rational.

%---------------------------------------------------------------------------%

:- type tower_elem
    --->    t_rat(rational)
    ;       t_ext(
                te_coeffs :: list(tower_elem),
                te_level  :: tower_level
            ).

:- type tower_level
    --->    tower_level(
                tl_id       :: int,
                tl_degree   :: int,
                tl_root_deg :: int,
                tl_radicand :: tower_elem
            ).

    % Check if a tower element is zero.
    %
:- pred t_is_zero(tower_elem::in) is semidet.

    % Get the outermost tower level, or no for t_rat.
    %
:- func t_level(tower_elem) = maybe(tower_level).

    % Outermost level ID, or -1 for t_rat.
    %
:- func level_id(tower_elem) = int.

    % Promote to a given level by embedding as constant.
    %
:- func promote_to(tower_level, tower_elem) = tower_elem.

    % Adjoin an nth root: create a new level where α^n = r.
    % Returns (level, generator).
    %
:- func adjoin_tower_root(int, int, tower_elem) = {tower_level, tower_elem}.

    % Convert a tower element to a radical expression.
    %
:- func tower_to_rad_expr(tower_elem) = rad_expr(rational).

    % Tower element arithmetic.
    %
:- func t_add(tower_elem, tower_elem) = tower_elem.
:- func t_sub(tower_elem, tower_elem) = tower_elem.
:- func t_neg(tower_elem) = tower_elem.
:- func t_mul(tower_elem, tower_elem) = tower_elem.
:- func t_inv(tower_elem) = tower_elem.
:- func t_div(tower_elem, tower_elem) = tower_elem.
:- func t_from_int(int) = tower_elem.

    % Equality test.
    %
:- pred t_eq(tower_elem::in, tower_elem::in) is semidet.

%---------------------------------------------------------------------------%
%---------------------------------------------------------------------------%

:- implementation.

:- use_module int.
:- import_module require.
:- use_module string.

%---------------------------------------------------------------------------%
% Zero checking, level inspection
%---------------------------------------------------------------------------%

t_is_zero(t_rat(R)) :- integer.is_zero(numer(R)).
t_is_zero(t_ext(Cs, _)) :- list.all_true(t_is_zero, Cs).

t_level(t_rat(_)) = no.
t_level(t_ext(_, L)) = yes(L).

level_id(t_rat(_)) = -1.
level_id(t_ext(_, L)) = tl_id(L).

%---------------------------------------------------------------------------%
% Equality
%---------------------------------------------------------------------------%

:- pred levels_eq(tower_level::in, tower_level::in) is semidet.

levels_eq(A, B) :- tl_id(A) = tl_id(B).

t_eq(t_rat(A), t_rat(B)) :- A = B.
t_eq(t_ext(Cs1, L1), t_ext(Cs2, L2)) :-
    levels_eq(L1, L2),
    eq_coeffs(Cs1, Cs2).
t_eq(t_rat(A), t_ext(Cs, _)) :-
    eq_coeffs([t_rat(A)], Cs).
t_eq(t_ext(Cs, _), t_rat(A)) :-
    eq_coeffs(Cs, [t_rat(A)]).

:- pred eq_coeffs(list(tower_elem)::in, list(tower_elem)::in) is semidet.

eq_coeffs([], []).
eq_coeffs([], Bs) :- list.all_true(t_is_zero, Bs).
eq_coeffs(As, []) :- As = [_ | _], list.all_true(t_is_zero, As).
eq_coeffs([A | As], [B | Bs]) :- t_eq(A, B), eq_coeffs(As, Bs).

%---------------------------------------------------------------------------%
% Trimming and smart constructor
%---------------------------------------------------------------------------%

:- func trim_te(list(tower_elem)) = list(tower_elem).

trim_te(Cs) = list.reverse(drop_while_zero(list.reverse(Cs))).

:- func drop_while_zero(list(tower_elem)) = list(tower_elem).

drop_while_zero(Xs) = Result :-
    ( if Xs = [H | T], t_is_zero(H) then
        Result = drop_while_zero(T)
    else
        Result = Xs
    ).

:- func mk_t_ext(list(tower_elem), tower_level) = tower_elem.

mk_t_ext(Cs, L) = Result :-
    Trimmed = trim_te(Cs),
    ( if Trimmed = [] then
        Result = t_rat(rational.zero)
    else if Trimmed = [C], C = t_rat(_) then
        Result = C
    else
        Result = t_ext(pad_to(tl_degree(L), Trimmed), L)
    ).

:- func pad_to(int, list(tower_elem)) = list(tower_elem).

pad_to(N, Cs) = Result :-
    Len = list.length(Cs),
    ( if int.'>='(Len, N) then
        Result = list.take_upto(N, Cs)
    else
        Result = Cs ++ list.duplicate(int.'-'(N, Len), t_rat(rational.zero))
    ).

%---------------------------------------------------------------------------%
% Promotion
%---------------------------------------------------------------------------%

promote_to(Lvl, E) = Result :-
    ( if E = t_rat(_) then
        Result = t_ext(
            [E | list.duplicate(int.'-'(tl_degree(Lvl), 1),
                                t_rat(rational.zero))],
            Lvl)
    else if E = t_ext(_, L) then
        ( if levels_eq(L, Lvl) then
            Result = E
        else if int.'<'(tl_id(L), tl_id(Lvl)) then
            Result = t_ext(
                [E | list.duplicate(int.'-'(tl_degree(Lvl), 1),
                                    t_rat(rational.zero))],
                Lvl)
        else
            require.error("promote_to: cannot demote")
        )
    else
        Result = E
    ).

%---------------------------------------------------------------------------%
% Arithmetic
%---------------------------------------------------------------------------%

t_add(A, B) = Result :-
    ( if A = t_rat(RA), B = t_rat(RB) then
        Result = t_rat(rational.'+'(RA, RB))
    else if A = t_rat(_), B = t_ext(Bs, L) then
        Result = t_ext(add_coeffs([A], Bs), L)
    else if A = t_ext(As, L), B = t_rat(_) then
        Result = t_ext(add_coeffs(As, [B]), L)
    else if A = t_ext(As, L1), B = t_ext(Bs, L2) then
        ( if levels_eq(L1, L2) then
            Result = mk_t_ext(add_coeffs(As, Bs), L1)
        else if int.'<'(tl_id(L1), tl_id(L2)) then
            Result = t_add(promote_to(L2, A), B)
        else
            Result = t_add(A, promote_to(L1, B))
        )
    else
        Result = t_rat(rational.zero)
    ).

t_sub(A, B) = t_add(A, t_neg(B)).

t_neg(E) = Result :-
    ( if E = t_rat(R) then
        Result = t_rat(rational.'-'(rational.zero, R))
    else if E = t_ext(Cs, L) then
        Result = t_ext(list.map(t_neg, Cs), L)
    else
        Result = E
    ).

t_mul(A, B) = Result :-
    ( if A = t_rat(RA), B = t_rat(RB) then
        Result = t_rat(rational.'*'(RA, RB))
    else if A = t_rat(_), B = t_ext(Bs, L) then
        Result = mk_t_ext(list.map(( func(X) = t_mul(A, X) ), Bs), L)
    else if A = t_ext(As, L), B = t_rat(_) then
        Result = mk_t_ext(list.map(( func(X) = t_mul(X, B) ), As), L)
    else if A = t_ext(As, L1), B = t_ext(Bs, L2) then
        ( if levels_eq(L1, L2) then
            Result = mk_t_ext(
                reduce_coeffs(L1, poly_mul_te(As, Bs)), L1)
        else if int.'<'(tl_id(L1), tl_id(L2)) then
            Result = t_mul(promote_to(L2, A), B)
        else
            Result = t_mul(A, promote_to(L1, B))
        )
    else
        Result = t_rat(rational.zero)
    ).

t_inv(E) = Result :-
    ( if t_is_zero(E) then
        require.error("t_inv: division by zero")
    else if E = t_rat(R) then
        Result = t_rat(rational.'/'(rational.one, R))
    else if E = t_ext(Cs, L) then
        Result = t_inv_ext(Cs, L)
    else
        require.error("t_inv: impossible")
    ).

t_div(A, B) = t_mul(A, t_inv(B)).

t_from_int(N) = t_rat(rational(N, 1)).

%---------------------------------------------------------------------------%
% Coefficient-list arithmetic
%---------------------------------------------------------------------------%

:- func add_coeffs(list(tower_elem), list(tower_elem)) = list(tower_elem).

add_coeffs([], Bs) = Bs.
add_coeffs(As, []) = As :-
    As = [_ | _].
add_coeffs([A | As], [B | Bs]) = [t_add(A, B) | add_coeffs(As, Bs)].

:- func poly_mul_te(list(tower_elem), list(tower_elem))
    = list(tower_elem).

poly_mul_te(As, Bs) = Result :-
    ( if As = [] then
        Result = []
    else if Bs = [] then
        Result = []
    else
        RLen = int.'-'(int.'+'(list.length(As), list.length(Bs)), 1),
        Zeros = list.duplicate(RLen, t_rat(rational.zero)),
        Terms = list.condense(
            list.map(
                ( func({I, A}) =
                    list.map(
                        ( func({J, B}) = {int.'+'(I, J), t_mul(A, B)} ),
                        indexed(Bs))
                ),
                indexed(As))),
        Result = list.foldl(
            ( func({Idx, C}, Acc) = add_at(Idx, C, Acc) ),
            Terms, Zeros)
    ).

:- func indexed(list(tower_elem)) = list({int, tower_elem}).

indexed(Xs) = indexed_from(0, Xs).

:- func indexed_from(int, list(tower_elem)) = list({int, tower_elem}).

indexed_from(_, []) = [].
indexed_from(I, [X | Xs]) = [{I, X} | indexed_from(int.'+'(I, 1), Xs)].

:- func add_at(int, tower_elem, list(tower_elem)) = list(tower_elem).

add_at(_, _, []) = [].
add_at(I, C, [X | Xs]) = Result :-
    ( if I = 0 then
        Result = [t_add(X, C) | Xs]
    else
        Result = [X | add_at(int.'-'(I, 1), C, Xs)]
    ).

:- func reduce_coeffs(tower_level, list(tower_elem)) = list(tower_elem).

reduce_coeffs(L, Cs) = Result :-
    N = tl_degree(L),
    ( if int.'=<'(list.length(Cs), N) then
        Result = Cs
    else
        Lo = list.take_upto(N, Cs),
        Hi = drop_n(N, Cs),
        Shifted = list.map(
            ( func(X) = t_mul(X, tl_radicand(L)) ), Hi),
        Combined = add_coeffs(Lo, Shifted),
        Result = reduce_coeffs(L, Combined)
    ).

:- func drop_n(int, list(T)) = list(T).

drop_n(N, Xs) = Result :-
    ( if int.'=<'(N, 0) then
        Result = Xs
    else if Xs = [_ | Tail] then
        Result = drop_n(int.'-'(N, 1), Tail)
    else
        Result = []
    ).

%---------------------------------------------------------------------------%
% Inversion via extended GCD
%---------------------------------------------------------------------------%

:- func t_inv_ext(list(tower_elem), tower_level) = tower_elem.

t_inv_ext(Cs, L) = Result :-
    ( if list.all_true(t_is_zero, Cs) then
        require.error("t_inv_ext: division by zero")
    else
        MinP = min_poly_coeffs(L),
        {_G, S, _T} = poly_ext_gcd(trim_te(Cs), MinP),
        Reduced = reduce_coeffs(L, S),
        Result = mk_t_ext(Reduced, L)
    ).

:- func min_poly_coeffs(tower_level) = list(tower_elem).

min_poly_coeffs(L) = Result :-
    NegR = t_neg(tl_radicand(L)),
    Zeros = list.duplicate(int.'-'(tl_degree(L), 1), t_rat(rational.zero)),
    Result = [NegR | Zeros] ++ [t_rat(rational.one)].

:- func poly_ext_gcd(list(tower_elem), list(tower_elem))
    = {list(tower_elem), list(tower_elem), list(tower_elem)}.

poly_ext_gcd(A, B) =
    poly_ext_gcd_loop(A, B,
        [t_rat(rational.one)], [],
        [], [t_rat(rational.one)]).

:- func poly_ext_gcd_loop(
    list(tower_elem), list(tower_elem),
    list(tower_elem), list(tower_elem),
    list(tower_elem), list(tower_elem))
    = {list(tower_elem), list(tower_elem), list(tower_elem)}.

poly_ext_gcd_loop(R0, R1, S0, S1, T0, T1) = Result :-
    ( if list.all_true(t_is_zero, R1) then
        LC = last_nonzero(R0),
        LCInv = t_inv(LC),
        R0P = list.map(( func(X) = t_mul(X, LCInv) ), R0),
        S0P = list.map(( func(X) = t_mul(X, LCInv) ), S0),
        T0P = list.map(( func(X) = t_mul(X, LCInv) ), T0),
        Result = {trim_te(R0P), trim_te(S0P), trim_te(T0P)}
    else
        {Q, R} = poly_div_mod(R0, R1),
        S2 = poly_sub(S0, poly_mul_te(Q, S1)),
        T2 = poly_sub(T0, poly_mul_te(Q, T1)),
        Result = poly_ext_gcd_loop(R1, R, S1, S2, T1, T2)
    ).

:- func last_nonzero(list(tower_elem)) = tower_elem.

last_nonzero(Xs) = Result :-
    Trimmed = trim_te(Xs),
    ( if Trimmed = [] then
        Result = t_rat(rational.one)
    else
        Result = list.det_last(Trimmed)
    ).

:- func poly_div_mod(list(tower_elem), list(tower_elem))
    = {list(tower_elem), list(tower_elem)}.

poly_div_mod(F, G) = Result :-
    F1 = trim_te(F),
    G1 = trim_te(G),
    DegF = int.'-'(list.length(F1), 1),
    DegG = int.'-'(list.length(G1), 1),
    ( if int.'<'(DegF, DegG) then
        Result = {[], F1}
    else
        LCG = list.det_last(G1),
        Result = poly_div_loop([], F1, G1, DegG, LCG)
    ).

:- func poly_div_loop(list(tower_elem), list(tower_elem),
    list(tower_elem), int, tower_elem)
    = {list(tower_elem), list(tower_elem)}.

poly_div_loop(Q, R, G, DegG, LCG) = Result :-
    R1 = trim_te(R),
    DegR = int.'-'(list.length(R1), 1),
    ( if int.'<'(DegR, DegG) then
        Result = {trim_te(Q), R1}
    else if list.all_true(t_is_zero, R1) then
        Result = {trim_te(Q), R1}
    else
        LCR = list.det_last(R1),
        C = t_div(LCR, LCG),
        D = int.'-'(DegR, DegG),
        Term = list.duplicate(D, t_rat(rational.zero)) ++ [C],
        R2 = poly_sub(R1, poly_mul_te(Term, G)),
        Result = poly_div_loop(add_coeffs(Q, Term), R2, G, DegG, LCG)
    ).

:- func poly_sub(list(tower_elem), list(tower_elem)) = list(tower_elem).

poly_sub([], Bs) = list.map(t_neg, Bs).
poly_sub(As, []) = As :- As = [_ | _].
poly_sub([A | As], [B | Bs]) = [t_sub(A, B) | poly_sub(As, Bs)].

%---------------------------------------------------------------------------%
% Tower construction
%---------------------------------------------------------------------------%

adjoin_tower_root(LvlId, N, R) = {Level, Gen} :-
    Level = tower_level(LvlId, N, N, R),
    Gen = t_ext(
        [t_rat(rational.zero), t_rat(rational.one) |
         list.duplicate(int.'-'(N, 2), t_rat(rational.zero))],
        Level).

%---------------------------------------------------------------------------%
% Conversion to RadExpr
%---------------------------------------------------------------------------%

tower_to_rad_expr(t_rat(R)) = re_lit(R).
tower_to_rad_expr(t_ext(Cs, Level)) = Result :-
    Gen = re_root(tl_root_deg(Level),
                  tower_to_rad_expr(tl_radicand(Level))),
    Terms = make_terms(Cs, Gen, 0),
    ( if Terms = [] then
        Result = re_lit(rational.zero)
    else if Terms = [T] then
        Result = T
    else
        Result = list.foldl(
            ( func(X, Acc) = re_add(Acc, X) ),
            list.det_tail(Terms),
            list.det_head(Terms))
    ).

:- func make_terms(list(tower_elem), rad_expr(rational), int)
    = list(rad_expr(rational)).

make_terms([], _, _) = [].
make_terms([C | Cs], Gen, I) = Result :-
    ( if t_is_zero(C) then
        Result = make_terms(Cs, Gen, int.'+'(I, 1))
    else
        CE = tower_to_rad_expr(C),
        ( if I = 0 then
            Term = CE
        else if I = 1 then
            Term = re_mul(CE, Gen)
        else
            Term = re_mul(CE, re_pow(Gen, I))
        ),
        Result = [Term | make_terms(Cs, Gen, int.'+'(I, 1))]
    ).

%---------------------------------------------------------------------------%
:- end_module dyn_tower.
%---------------------------------------------------------------------------%

File: haskell/src/Surd/Field/DynTower.hs
-- | Dynamically-nested algebraic extension tower.
--
-- Unlike 'ExtElem' which uses Haskell's type system for tower nesting
-- (requiring static depth), 'TowerElem' uses a recursive data type
-- that supports towers of arbitrary runtime-determined depth.
--
-- This enables the Gauss period descent to build field towers
-- incrementally, one extension at a time, without knowing the
-- final depth at compile time.
--
-- Elements support full field arithmetic (Num, Fractional) and
-- can be converted to radical expressions for display.
module Surd.Field.DynTower
  ( TowerElem (..),
    TowerLevel (..),
    adjoinTowerRoot,
    towerToRadExpr,
    tIsZero,
    tLevel,
    promoteTo,
    levelId,
  )
where

import Surd.Types (RadExpr (..))

-- | Element of a dynamically-nested field tower.
--
-- @TRat r@ is an element of Q.
-- @TExt [a₀, a₁, ..., a_{d-1}] level@ represents a₀ + a₁·α + ... + a_{d-1}·α^{d-1}
-- where α is the generator of the extension described by @level@.
-- Coefficients are themselves 'TowerElem' values at the previous level.
data TowerElem
  = TRat !Rational
  | TExt ![TowerElem] !TowerLevel
  deriving (Show)

-- | One level of the field tower, describing a single simple radical
-- extension \(K(\alpha)\) where \(\alpha^n = r\) for some element
-- \(r\) of the base field \(K\).
data TowerLevel = TowerLevel
  { -- | Unique level identifier (assigned by caller)
    tlId :: !Int,
    -- | Extension degree (= @tlRootDeg@ for radical extensions)
    tlDegree :: !Int,
    -- | Root degree \(n\): the generator satisfies \(\alpha^n = r\)
    tlRootDeg :: !Int,
    -- | Radicand \(r\): the generator is \(\sqrt[n]{r}\)
    tlRadicand :: !TowerElem
  }
  deriving (Show)

instance Eq TowerLevel where
  a == b = tlId a == tlId b

-- ---------------------------------------------------------------------------
-- Zero checking and level inspection
-- ---------------------------------------------------------------------------

-- | Test whether a tower element is zero.
--
-- For rational elements, checks @r == 0@. For extension elements,
-- checks that all coefficients are recursively zero. This is needed
-- for coefficient trimming, equality testing, and zero-divisor checks
-- in the extended GCD inversion algorithm.
tIsZero :: TowerElem -> Bool
tIsZero (TRat r) = r == 0
tIsZero (TExt cs _) = all tIsZero cs

-- | Get the outermost extension level of a tower element, or
-- 'Nothing' for a rational element (which lives in the base field Q).
tLevel :: TowerElem -> Maybe TowerLevel
tLevel (TRat _) = Nothing
tLevel (TExt _ l) = Just l

-- ---------------------------------------------------------------------------
-- Eq instance
-- ---------------------------------------------------------------------------

instance Eq TowerElem where
  TRat a == TRat b = a == b
  TExt cs1 l1 == TExt cs2 l2
    | l1 == l2 = eqCoeffs cs1 cs2
    | otherwise = False
  TRat a == TExt cs _ = eqCoeffs [TRat a] cs
  TExt cs _ == TRat a = eqCoeffs cs [TRat a]

-- | Compare coefficient lists, treating missing trailing entries as zero.
eqCoeffs :: [TowerElem] -> [TowerElem] -> Bool
eqCoeffs [] [] = True
eqCoeffs [] bs = all tIsZero bs
eqCoeffs as [] = all tIsZero as
eqCoeffs (a : as) (b : bs) = a == b && eqCoeffs as bs

-- ---------------------------------------------------------------------------
-- Trimming
-- ---------------------------------------------------------------------------

-- | Remove trailing zero coefficients.
trimTE :: [TowerElem] -> [TowerElem]
trimTE = reverse . dropWhile tIsZero . reverse

-- ---------------------------------------------------------------------------
-- Num instance
-- ---------------------------------------------------------------------------

-- | Promote a tower element to a given level by embedding as the
-- constant coefficient in the higher extension.
--
-- If element @e@ lives in field @K@, then @promoteTo lvl e@ embeds it
-- as @e + 0*α + 0*α² + ...@ in @K(α)@ where @α@ is the generator
-- of @lvl@. Errors if the element already lives at a /higher/ level
-- than the target (demotion is not supported).
promoteTo :: TowerLevel -> TowerElem -> TowerElem
promoteTo lvl e@(TRat _) = TExt (e : replicate (tlDegree lvl - 1) (TRat 0)) lvl
promoteTo lvl e@(TExt _ l)
  | l == lvl = e
  | tlId l < tlId lvl = TExt (e : replicate (tlDegree lvl - 1) (TRat 0)) lvl
  | otherwise = error $ "promoteTo: cannot demote level " ++ show (tlId l) ++ " to " ++ show (tlId lvl)

-- | Get the outermost level ID, or @-1@ for 'TRat' (base field Q).
--
-- Level IDs are assigned by the caller of 'adjoinTowerRoot' and should
-- be globally unique within a tower construction. Higher IDs correspond
-- to later (outer) extensions.
levelId :: TowerElem -> Int
levelId (TRat _) = -1
levelId (TExt _ l) = tlId l

instance Num TowerElem where
  -- Addition
  TRat a + TRat b = TRat (a + b)
  TRat a + TExt bs l = TExt (addCoeffs [TRat a] bs) l
  TExt as l + TRat b = TExt (addCoeffs as [TRat b]) l
  TExt as l1 + TExt bs l2
    | l1 == l2 = mkTExt (addCoeffs as bs) l1
    | tlId l1 < tlId l2 = promoteTo l2 (TExt as l1) + TExt bs l2
    | otherwise = TExt as l1 + promoteTo l1 (TExt bs l2)

  -- Multiplication
  TRat a * TRat b = TRat (a * b)
  TRat a * TExt bs l = mkTExt (map (TRat a *) bs) l
  TExt as l * TRat b = mkTExt (map (* TRat b) as) l
  TExt as l1 * TExt bs l2
    | l1 == l2 = mkTExt (reduceCoeffs l1 (polyMulTE as bs)) l1
    | tlId l1 < tlId l2 = promoteTo l2 (TExt as l1) * TExt bs l2
    | otherwise = TExt as l1 * promoteTo l1 (TExt bs l2)

  negate (TRat a) = TRat (negate a)
  negate (TExt cs l) = TExt (map negate cs) l

  abs = error "TowerElem: abs not meaningful"
  signum = error "TowerElem: signum not meaningful"
  fromInteger n = TRat (fromInteger n)

-- ---------------------------------------------------------------------------
-- Fractional instance
-- ---------------------------------------------------------------------------

instance Fractional TowerElem where
  recip (TRat a) = TRat (recip a)
  recip (TExt cs l) = tInv cs l

  fromRational = TRat

-- | Smart constructor: if all coefficients are zero, produce TRat 0.
mkTExt :: [TowerElem] -> TowerLevel -> TowerElem
mkTExt cs l =
  let cs' = trimTE cs
   in case cs' of
        [] -> TRat 0
        [c] | isBaseLevel c -> c
        _ -> TExt (padTo (tlDegree l) cs') l
  where
    isBaseLevel (TRat _) = True
    isBaseLevel _ = False

-- | Pad a coefficient list to exactly n elements with zeros.
padTo :: Int -> [TowerElem] -> [TowerElem]
padTo n cs
  | length cs >= n = take n cs
  | otherwise = cs ++ replicate (n - length cs) (TRat 0)

-- ---------------------------------------------------------------------------
-- Coefficient-list arithmetic
-- ---------------------------------------------------------------------------

addCoeffs :: [TowerElem] -> [TowerElem] -> [TowerElem]
addCoeffs [] bs = bs
addCoeffs as [] = as
addCoeffs (a : as) (b : bs) = (a + b) : addCoeffs as bs

-- | Schoolbook polynomial multiplication on coefficient lists.
polyMulTE :: [TowerElem] -> [TowerElem] -> [TowerElem]
polyMulTE [] _ = []
polyMulTE _ [] = []
polyMulTE as bs =
  let rlen = length as + length bs - 1
      zeros = replicate rlen (TRat 0)
      terms =
        [ (i + j, a * b)
          | (i, a) <- zip [0 :: Int ..] as,
            (j, b) <- zip [0 :: Int ..] bs
        ]
   in foldl (\acc (idx, c) -> addAt idx c acc) zeros terms
  where
    addAt _ _ [] = []
    addAt 0 c (x : xs) = (x + c) : xs
    addAt i c (x : xs) = x : addAt (i - 1) c xs

-- | Reduce a coefficient list modulo α^n = r.
-- For a binomial minimal polynomial x^n - r, reduction is simple:
-- replace α^n with r, α^{n+1} with r·α, etc.
reduceCoeffs :: TowerLevel -> [TowerElem] -> [TowerElem]
reduceCoeffs l cs
  | length cs <= n = cs
  | otherwise =
      let (lo, hi) = splitAt n cs
          -- α^{n+k} = r · α^k
          shifted = map (* tlRadicand l) hi
          combined = addCoeffs lo shifted
       in reduceCoeffs l combined
  where
    n = tlDegree l

-- ---------------------------------------------------------------------------
-- Inversion via extended GCD
-- ---------------------------------------------------------------------------

-- | Invert an element in an extension field.
-- Uses extended GCD: find s such that s·elem ≡ 1 (mod minpoly).
tInv :: [TowerElem] -> TowerLevel -> TowerElem
tInv cs l
  | all tIsZero cs = error "TowerElem: division by zero"
  | otherwise =
      let minP = minPolyCoeffs l
          -- extGcd gives (g, s, t) with g = s·cs + t·minP
          (_g, s, _t) = polyExtGcd (trimTE cs) minP
          -- s is the inverse, reduce mod minpoly
          reduced = reduceCoeffs l s
       in -- Make monic: divide by leading coefficient of g
          mkTExt reduced l

-- | Build the minimal polynomial coefficient list for a tower level.
-- For α^n = r, the minimal polynomial is x^n - r = [-r, 0, ..., 0, 1].
minPolyCoeffs :: TowerLevel -> [TowerElem]
minPolyCoeffs l =
  negate (tlRadicand l) : replicate (tlDegree l - 1) (TRat 0) ++ [TRat 1]

-- | Extended GCD for coefficient lists (polynomials over TowerElem).
-- Returns (g, s, t) such that g = s·a + t·b.
polyExtGcd :: [TowerElem] -> [TowerElem] -> ([TowerElem], [TowerElem], [TowerElem])
polyExtGcd a b = go a b [TRat 1] [] [] [TRat 1]
  where
    go r0 r1 s0 s1 t0 t1
      | all tIsZero r1 =
          -- Make monic
          let lc = lastNonZero r0
              r0' = map (/ lc) r0
              s0' = map (/ lc) s0
              t0' = map (/ lc) t0
           in (trimTE r0', trimTE s0', trimTE t0')
      | otherwise =
          let (q, r) = polyDivMod r0 r1
              s2 = polySub s0 (polyMulTE q s1)
              t2 = polySub t0 (polyMulTE q t1)
           in go r1 r s1 s2 t1 t2

    lastNonZero xs = case trimTE xs of
      [] -> TRat 1
      ts -> last ts

-- | Polynomial division with remainder on coefficient lists.
polyDivMod :: [TowerElem] -> [TowerElem] -> ([TowerElem], [TowerElem])
polyDivMod _ [] = error "polyDivMod: division by zero"
polyDivMod f g
  | degF < degG = ([], f)
  | otherwise = go [] f
  where
    f' = trimTE f
    g' = trimTE g
    degF = length f' - 1
    degG = length g' - 1
    lcG = last g'

    go q r =
      let r' = trimTE r
          degR = length r' - 1
       in if degR < degG || all tIsZero r'
            then (trimTE q, r')
            else
              let lcR = last r'
                  c = lcR / lcG
                  d = degR - degG
                  -- c * x^d
                  term = replicate d (TRat 0) ++ [c]
                  r'' = polySub r' (polyMulTE term g')
               in go (addCoeffs q term) r''

-- | Polynomial subtraction on coefficient lists.
polySub :: [TowerElem] -> [TowerElem] -> [TowerElem]
polySub [] bs = map negate bs
polySub as [] = as
polySub (a : as) (b : bs) = (a - b) : polySub as bs

-- ---------------------------------------------------------------------------
-- Tower construction
-- ---------------------------------------------------------------------------

-- | Adjoin an \(n\)th root to the tower: create a new extension level
-- where the generator \(\alpha\) satisfies \(\alpha^n = r\).
--
-- Returns the new 'TowerLevel' and the generator element \(\alpha\),
-- represented as @[0, 1, 0, ..., 0]@ (the polynomial \(\alpha\)) in
-- the new extension. The level ID must be globally unique within the
-- tower being constructed (the caller is responsible for this).
adjoinTowerRoot ::
  -- | Level ID (unique)
  Int ->
  -- | Root degree n
  Int ->
  -- | Radicand r (element of current field)
  TowerElem ->
  (TowerLevel, TowerElem)
adjoinTowerRoot lvlId n r =
  let level =
        TowerLevel
          { tlId = lvlId,
            tlDegree = n,
            tlRootDeg = n,
            tlRadicand = r
          }
      -- The generator α, represented as [0, 1, 0, ..., 0]
      gen = TExt (TRat 0 : TRat 1 : replicate (n - 2) (TRat 0)) level
   in (level, gen)

-- ---------------------------------------------------------------------------
-- Conversion to RadExpr
-- ---------------------------------------------------------------------------

-- | Convert a tower element to a radical expression ('RadExpr').
--
-- Recursively expands each extension level's generator \(\alpha\) as
-- @'Root' n (towerToRadExpr r)@ where \(\alpha^n = r\), and
-- reconstructs the polynomial expression
-- \(c_0 + c_1 \alpha + \cdots + c_{d-1} \alpha^{d-1}\).
--
-- Warning: the resulting expression may be exponentially large for deep
-- towers, since each level's generator is expanded independently. For
-- display purposes, prefer the structured rendering in
-- "Surd.Field.DynTower.Display".
towerToRadExpr :: TowerElem -> RadExpr Rational
towerToRadExpr (TRat r) = Lit r
towerToRadExpr (TExt cs level) =
  let gen = Root (tlRootDeg level) (towerToRadExpr (tlRadicand level))
      terms =
        [ case i of
            0 -> towerToRadExpr c
            1 -> Mul (towerToRadExpr c) gen
            _ -> Mul (towerToRadExpr c) (Pow gen i)
          | (c, i) <- zip cs [0 :: Int ..],
            not (tIsZero c)
        ]
   in case terms of
        [] -> Lit 0
        [t] -> t
        _ -> foldl1 Add terms
